### ISSUE TITLE
addEventListener is not defined

### ISSUE THREAD

**Issue by jguillen1984** [layerhq/layer-websdk#151]
When upgrading to layer-websdk 3.2.1 from 3.0.0 I started to get this error on my builds.
```
ReferenceError: addEventListener is not defined
    at Object.<anonymous> (/Users/johnny.guillen/dev/projects/hac-messenger/node_modules/layer-websdk/lib/client-utils.js:189:1)
```

it looks like if failing on this block of code:
```
addEventListener('message', function (event) {
  if (event.data.type !== 'layer-set-immediate') return;
  setImmediateProcessor();
});
```

It looks like our webpack bundler fails whenever we encounter this line.

@michaelkantor any ideas?

**Comment by benhorst** [comment]
https://github.com/layerhq/layer-websdk/blob/master/src/client-utils.js#L176

**Comment by michaelkantor** [comment]
Is this a webpack only issue, or are you building a React Native build?

If its React Native, we have fixes coming for that in our next patch release.

My webpack build  does not show this error; can you try and replace `addEventListener` with `global. addEventListener` and see if that makes the problem go away?  Not really clear what the issue here is if its not React Native related.

Michael

**Comment by benhorst** [comment]
@michaelkantor it's a webpack build for browser. If it isn't happening for you, I have a bad feeling it might have to do with "use strict" being defined somewhere in our build in which case, the global/window object wouldn't be inferred. I know we tried adding `document.addEventListener` earlier today and it was ok at build-time (but I imagine that would fail at runtime because the event isn't likely fired on document). Will check in with further info soon.

**Comment by michaelkantor** [comment]
`global === window`; these are the same (except in react native)

So just try `global.addEventListener` and see if it compiles.  If so, I'll run some validation on my end and ship it.  

I'm trying to ship an update tonight, but will wait for mid morning tomorrow so you can report back.  Otherwise, this fix will wait for the patch release after next.

Michael

**Comment by benhorst** [comment]
Sure, I understand they are the same concept, but this is a webpack build targeting browser, AFAIK the `global` object exists in the Node runtime, but browser side is represented by `window`. `global` is going to be undefined in browsers. @jguillen1984 is checking it out.

**Comment by michaelkantor** [comment]
Just to clarify: `global` is defined by browserify/webpack within the context of your module definition.  Your module gets wrapped in a function at build time, and `global` is inserted into that function scoped to be visible to all code within your module.

**Comment by jguillen1984** [comment]
@michaelkantor - So we are unable to reproduce the issue using a simple webpack bundle. We are investigating the issue further on our end to determine what in our build pipeline could be causing this.

**Comment by jguillen1984** [comment]
_Update_
Still haven't found an issue in our build pipeline. Continuing to dig deeper.

I was able to get it to compile by wrapping the code with `if (window) { addEventListener(.... }`.  I also attempted to use `global.addEventListener` and `window.addEventListener`, but both failed for me.

A side note is that after trying 3.2.3 it now fails on `https://github.com/layerhq/layer-websdk/blob/master/src/client-utils.js#L9`. I tried both `import layer from 'layer-websdk'` and `import layer from 'layer-websdk/index-es6'` with the same results.

We are running currently `node 6.9.2` and `webpack 2.5.0` with `webpack-dev-server`

**Comment by michaelkantor** [comment]
We'll continue to iterate on this.  But there's a lot of guesswork involved.  `global.addEventListener` should work, can you show the errors you get from it?

**Comment by jguillen1984** [comment]
```
/Users/johnny.guillen/dev/projects/hac-messenger/node_modules/layer-websdk/lib/client-utils.js:189
global.addEventListener('message', function (event) {
       ^

TypeError: global.addEventListener is not a function
    at Object.<anonymous> (/Users/johnny.guillen/dev/projects/hac-messenger/node_modules/layer-websdk/lib/client-utils.js:189:8)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Function.cls_wrapMethod [as _load] (/Users/johnny.guillen/dev/projects/hac-messenger/node_modules/newrelic/lib/shimmer.js:256:38)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/johnny.guillen/dev/projects/hac-messenger/node_modules/layer-websdk/lib/root.js:15:13)
    at Module._compile (module.js:570:32)
5 May 15:20:18 - [nodemon] app crashed - waiting for file changes before starting..
```

**Comment by michaelkantor** [comment]
Ok. I would like to resolve this. But the back and forth on this is liable to kill us...

Can you setup a barebones github or zip project that replicates this?  I expect this would ultimately be easier on both of us.  If not, we'll keep going as-is.  For the problems your seeing, you can use fake app-ids, and leave out any kind of credentials, UI, etc...  We just need something that gets built not run (and which fails to get built).

**Comment by benhorst** [comment]
Totally agree. We will attempt to produce a minimal repro. Since it doesn't occur in a naive sample, we wouldn't expect you to change anything until we know what is wrong.

**Comment by michaelkantor** [comment]
Closing this for now.  Reopen this if there's anything I can do to help.

**Comment by jguillen1984** [comment]
@michaelkantor - ok so I was able to get this to work for our build. 

I upgraded layer-websdk to v3.3.2, then I tried to build my project and got an error on `global.getNativeSupport('atob') is undefined` https://github.com/layerhq/layer-websdk/blob/master/src/client-utils.js#L9

In node the library atob does not exist so it was returning "undefined", which in turn tried to call `global.getNativeSupport` and that function was not part of the global scope.  I then noticed that you added this to your react-native build in the global scope. So I dug into node_modules folder for layer-websdk and did the same for the `index.js` file and it all works :tada:.

here's what that looks like now:
```
/* istanbul ignore next */
if (global.layer && global.layer.Client) {
  console.error('ERROR: It appears that you have multiple copies of the Layer WebSDK in your build!');
} else {
  global.getNativeSupport = function(module) {
      switch(module) {
        case 'atob':
          return function(b64Str) {
            let byteArray = base64JS.toByteArray(b64Str);
            let strArray = [];
            for (let i=0, l=byteArray.length; i<l; i++) {
              strArray[i] = String.fromCharCode(byteArray[i]);
            }
            return strArray.join('');
          }
        case 'btoa':
          return function(str) {
            let arr = str.split("").map((val) => {
              return val.charCodeAt(0);
            });
            return base64JS.fromByteArray(arr);
          }
        case 'setImmediate':
          // Globally defined by RN environment
          return setImmediate;
        case 'OnlineEvents':
          return NetInfo;
      }
  };
  global.layer = require('./lib/layer');
}
module.exports = global.layer;
```
